core/src/errors.rs

```diff
@@ -3,7 +3,7 @@ use core::fmt;
 // INPUT ERROR
 // ================================================================================================
 
-#[derive(Clone, Debug)]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
```

Comment:
A couple of questions (which are relevant to other places as well):
- I think adding things like `Eq` and `PartialEq` is fine, but what is the motivation for adding `Hash`?
- Do we need to add `Copy` here?

Reply:
You should always add `Copy` for types that are trivially copyable IMO, the only exception are types that implement `Drop`, but I'm not sure you can even do that (it's non-sensical). It's a marker trait, and strictly an optimization for cloneable types, so there is no downside that i can think of.

No idea if it applies here, but `Hash` is commonly derived for use with `HashMap`, either for use directly as a key or because many of its APIs can take keys which are `K: Borrow<T>`, where `T: Hash + Eq`, as long as the types hash to the same value.

Reply:
Looking closer at the actual code though, I agree with @bobbinth, error types usually shouldn't implement Copy because it makes extending the enumeration a breaking change in some cases, and Hash seems out of place on an error type.

Reply:
I'm following the recommended guidelines:

https://rust-lang.github.io/api-guidelines/interoperability.html#types-eagerly-implement-common-traits-c-common-traits

Note: For this PR I did the simple ones that are just a derive, eventually we should also have `Display`, `Default`, and the conversion traits.

---

As for the question, each trait has a different use case. Here is an overview of their use cases and my opinion on it:

Should **always**, with no exception, be derived:

- `Debug`: used for simple `dbg!` debugging
- `Clone`: necessary in some cases because for safety (i.e. functions that work on references but need an owning copy)
- `PartialEq`: At the very minimum we need this for testing. It is actually surprising to me that so many structs did not have this implemented, meaning we are lacking a lot of simple unit tests in the codebase.
- `Hash`: Enables using hash tables. This needs to be implemented from the bottom-up.
- `Display`: To have useful user messages. This also needs to be implemented from the bottom-up.

Note: This PR does not add the above to all cases, eventually the gap should be fixed (most likely some subtype didn't implement the above traits in another repo/crate).

When possible should **always**, be derived:

- `Copy`: As per @bitwalker explanation
- `Default`: some cases it makes no sense to have a default, but if it makes sense, then it allows for cleaner code, e.g. [unwrap_or_default](https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap_or_default) for error handling.
- `Eq`, `PartialOrd`, `Ord`: Enables functionality, specially using things like b-trees

---

I have no idea why error should be treated differently and not follow the above, I personally don't see any downsides, a very simple reason to have errors implement hash is that we can add errors to containers, for instance, on a server that wants to count errors it has seen. We can also use it to dedup errors on retries. It only makes it backwards incompatible if a new error variant is introduced which contains a sub-type that itself doesn't implement `Hash`, but if we follow the guidelines that shouldn't happen, i.e. it is a bug in the subtype and not in the error.

Reply:
Largely agree with all of that. The reason why Copy is different for errors is because if you add a variant that can't be Copy, then it breaks all downstream users that rely on the error type being Copy. Strictly speaking this applies to any public type that you might extend in such a way, namely enums. Copy is tricky in general because of how downstream code changes based on things being Copy vs Clone, even for non-public types (though the impact is far less significant in that case).

I see little reason to implement Hash for error types though, unless you specifically plan to use them as HashMap keys, which doesn't make sense anyway.

You can also get away without PartialEq in many cases if it's just used for testing by using `assert_matches`, particularly for error types which might not have a meaningful notion of equality.

My advice is to never blindly derive traits unless the trait is meaningful for the type. Especially if you derive traits that are never being used. The calculus is different for public types, to some degree, but I think it still ultimately comes back to whether the trait has a meaningful implementation or not, especially for things like PartialEq and PartialOrd/Ord. Anyway, I still think following the recommended guidelines as a default is better than not implementing any traits, I just think the topic is more nuanced. The guidelines are meant to encourage the more useful default for those who aren't sure what to implement for their types, rather than dictating what should be implemented. But that's just my take, I know it's not a universal opinion.

Reply:
> Copy is tricky in general because of how downstream code changes based on things being Copy vs Clone

Fair point.

You covered that a bit, but I just would like to extend that the issue is not with error types, but with forward compatibility of public types, that is to say, as you mentioned a non-error enums have the same issue, so do structs that gets a new field.

> I see little reason to implement Hash for error types though, unless you specifically plan to use them as HashMap keys, which doesn't make sense anyway.

I don't agree with this, errors should not be special. For example `HashMap<Error, usize>` is one of the use that makes sense to me, to count errors for metric collection on long running servers. Another reason is to use a `HashSet<Error>` on a retry loop, to collect distinct errors.

To extend on the above, I see `Error` types as recoverable errors, and long-running servers should be allowed to compare/store/serialize them.

> You can also get away without PartialEq in many cases if it's just used for testing by using assert_matches, particularly for error types which might not have a meaningful notion of equality.

I can imagine some cases were `PartialEq` can't be implemented, say, when using an opaque pointer from a c library. But to me, it seems these are the exceptions, instead of regular cases, and as a guideline it makes sense to me to basically always derive it.

> My advice is to never blindly derive traits unless the trait is meaningful for the type

I wasn't trying to advocate for that. I am however saying it is best to have the derives than not have them. Specially for something that is being used as a library (like these crates), and for some of the derives

Reply:
If we can agree on the derives that are basically no brainers on **our** context. I can close this PR, and open another one with those added first (say `Debug` and `Clone`). And we can do the others some other time.

Reply:
> If we can agree on the derives that are basically no brainers on our context. I can close this PR, and open another one with those added first (say Debug and Clone). And we can do the others some other time.

This is totally bikeshedding IMO, so definitely don't hold up the PR based on the discussion, and I don't mean to suggest that you have to take any action on this either. Totally your call on how you approach this as far as I'm concerned.

> I do think we may be disagreeing on two things:
>
> * Which derives should be done without much thought needed
> * What is the scope of use for error types

I think we probably have different ideas about how error types are/should be used, yes, and so we disagree a bit in that regard, but on both matters I'm not super opinionated as long as there is a reasonable justification for a given choice. I have opinions of course, but certainly willing to err on the side of more derives than less. 

One point I would like to make though is that traits implicitly broaden the API surface of a type - which isn't _really_ an issue for things that are internal to a project; but if you are exposing that type from a library for public consumption, your downstream users can/will rely on those trait implementations as part of the public interface. If you later determine you want to change a type in some way that makes it incompatible with a given trait, that's a breaking change. This has bitten me on a number of occasions with various of my open source projects, unintentional breaking changes is something people _absolutely rage_ about when it happens, and who can really blame them. As a result, I tend to be more conservative now when it comes to public APIs, but outside of that, I'm all for maximizing the ways in which you can use types. I do think that if you _can_ implement a lot of the recommended traits, you absolutely should, just with an eye towards future maintainability so you don't tie your hands too much accidentally.

> For the first point, I think that Debug, Clone, Hash, Display, and PartialEq are basically always implementable.

Agreed, with the caveat that this is only true if you choose to omit fields that don't already implement them (and which you do not control), or provide some kind of ad-hoc implementation for them. I run into types like this on a regular basis. Error types in particular, e.g. `anyhow::Error` has no sane implementation for `Clone` or `PartialEq` (not sure about `Hash`, but I don't think it implements that either). Trait objects in general have this issue, unless the trait happens to have all of those as supertraits. The `std::error::Error` trait doesn't require anything but `Debug` and `Display`, and many (if not most/all) of the standard library error types don't implement anything other than the bare minimum required by `std::error::Error`, if even that much.

> I don't agree with this, errors should not be special. For example HashMap<Error, usize> is one of the use that makes sense to me, to count errors for metric collection on long running servers. Another reason is to use a HashSet<Error> on a retry loop, to collect distinct errors.

So perhaps to explain my position a bit: in my various Rust-based compiler projects, which have been the largest and most complex Rust projects I've worked on thus far, and which do a lot of complex instrumentation and error tracking, I've never done either of those, let alone had an error type that implemented `Hash`, and it's rarely the case that I even implement `Clone`. I will sometimes implement `PartialEq` (and even `Eq`) if the error type supports it and I have a particular need for it, but that's pretty rare as well, since many error types will have data that has no notion of equality. 

In particular, in my experience good errors tend to carry data unique to the instance in which it was raised, to provide better information to the end user, which often makes it uncloneable, and frequently cannot be hashed or compared in any meaningful sense. In those instances where I want `PartialEq` (for example), I'll typically compare by variant, ignoring the payload entirely (unless it has a meaningful notion of equality). A common issue is propagating errors. If you define a custom error type for some operation that might encounter an I/O error, you will likely want to wrap+propagate that error by having a variant like `Io(std::io::Error)` - well that variant has no meaningful implementation of `Clone`, `Hash`, or `PartialEq`, because the `std::io::Error` type doesn't implement any of those, and that's just one common example.

In any case, I'm not suggesting that there aren't times where you might want or need to implement those traits for an error, and that it is somehow a bad idea, I don't think that's true at all. That said, such use cases strike me as very situational, and not the typical case. One need only look at Rust's standard library to see which side of the coin error types tend to fall on. Ultimately, how often do you clone error values? How often do you hash them? Comparing for equality happens a lot in tests perhaps, but outside of tests, how often are you comparing for equality vs matching on them?

Anyway, I don't have any issues with your approach by any means, so this is really just a matter of me having been nerd-sniped by GitHub notifications again 😆